### PR TITLE
chaincfg/blockchain: Parameterize more chain consts.

### PR DESCRIPTION
**Pull request by davecgh**
This moves several of the chain constants to the `Params` struct in the `chaincfg` package which is intended for that purpose.  This is mostly a backport of the same modifications made in `Decred` along with a few additional things cleaned up.

The following is an overview of the changes:

- Comment all fields in the `Params` struct definition
- Add locals to `BlockChain` instance for the calculated values based on the provided chain params
- Rename the following param fields:
  - `SubsidyHalvingInterval` -> `SubsidyReductionInterval`
  - `ResetMinDifficulty` -> `ReduceMinDifficulty`
- Add new Param fields:
  - `CoinbaseMaturity`
  - `TargetTimePerBlock`
  - `TargetTimespan`
  - `BlocksPerRetarget`
  - `RetargetAdjustmentFactor`
  - `MinDiffReductionTime`

### CODE REVIEW

**Comment by dajohi** [comment]
OK

File: rpcserver.go [hunk 1]
```diff
@@ -2104,6 +2104,11 @@ func handleGetNetworkHashPS(s *rpcServer, cmd interface{}, closeChan <-chan stru
 		endHeight = best.Height
 	}
 
+	// Calculate the number of blocks per retarget interval based on the
+	// chain parameters.
+	blocksPerRetarget := int32(s.server.chainParams.TargetTimespan /
+		s.server.chainParams.TargetTimePerBlock)
+
```
Comment (dajohi) [inline rpcserver.go]: Perhaps blockchain could have an BlocksPerRetarget() int32 { }  that returns what it was configured with.. instead of recalcuating here?

Comment (davecgh) [inline rpcserver.go]: It's based on constants, so it's probably not a big deal that it's calculating it here.  However, I don't imagine it would hurt to expose it on the chain instance either.

**Comment by cjepson** [comment]
OK. Goes a long way towards making these parameters more accessible by people who wish to alter them.

File: rpcserver.go (hunk 1 above)
Comment (dajohi) [inline rpcserver.go]: That can be a separate PR though.  OK on this PR.
